xdsl/dialects/builtin.py

```diff
@@ -183,14 +166,83 @@ def print_parameter(self, printer: Printer) -> None:
     @staticmethod
     def generic_constraint_coercion(args: tuple[Any]) -> AttrConstraint:
         assert len(args) == 1
-        return ArrayOfConstraint(irdl_to_attr_constraint(args[0]))
+        return ArrayOfConstraint(RangeOf(irdl_to_attr_constraint(args[0])))
 
     def __len__(self):
         return len(self.data)
 
     def __iter__(self) -> Iterator[AttributeCovT]:
         return iter(self.data)
 
+    @staticmethod
+    def constr(
+        constr: (
+            IRDLGenericAttrConstraint[AttributeCovT]
+            | GenericRangeConstraint[AttributeCovT]
+        ),
+    ) -> GenericAttrConstraint[ArrayAttr[AttributeCovT]]:
+        return ArrayOfConstraint(constr)
+
+
+@dataclass(frozen=True)
+class ArrayOfConstraint(GenericAttrConstraint[ArrayAttr[AttributeCovT]]):
+    elem_range_constraint: GenericRangeConstraint[AttributeCovT]
+    """
+    A constraint that enforces an ArrayData whose elements all satisfy
+    the elem_constr.
+    """
+
+    def __init__(
+        self,
+        constr: (
+            IRDLGenericAttrConstraint[Attribute] | GenericRangeConstraint[AttributeCovT]
+        ),
+    ):
+        if isinstance(constr, GenericRangeConstraint):
+            object.__setattr__(self, "elem_range_constraint", constr)
```

Comment:
why use `object.__setattr__`?

Reply:
frozen

Reply:
even within the init?

Reply:
Is there not `super().__init__(constr)` or something instead?

Reply:
not that I know of

Reply:
`super().__init__(constr)` wouldn't work, because `super()` will refer to the parent class.
This initialization thing for frozen class kinda sucks ngl